data/s3/src/main/java/org/teiid/spring/data/s3/S3Connection.java

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags and
+ * the COPYRIGHT.txt file distributed with this work.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.teiid.spring.data.s3;
+
+import com.amazonaws.SdkClientException;
+import com.amazonaws.services.s3.AmazonS3;
+import com.amazonaws.services.s3.model.ListObjectsRequest;
+import com.amazonaws.services.s3.model.ObjectListing;
+import com.amazonaws.services.s3.model.ObjectMetadata;
+import com.amazonaws.services.s3.model.S3ObjectSummary;
+import org.teiid.file.VirtualFile;
+import org.teiid.file.VirtualFileConnection;
+import org.teiid.translator.TranslatorException;
+
+import java.io.InputStream;
+import java.util.Vector;
+
+public class S3Connection implements VirtualFileConnection {
+
+
+    private final S3Configuration s3Config;
+    private final AmazonS3 s3Client;
+
+    public S3Connection(S3Configuration s3Config, AmazonS3 s3Client) {
+        this.s3Config = s3Config;
+        this.s3Client = s3Client;
+    }
+
+    @Override
+    public VirtualFile[] getFiles(String s) throws TranslatorException {
+        if(s3Client.doesObjectExist(s3Config.getBucket(), s)){
+            ListObjectsRequest listObjectsRequest = new ListObjectsRequest()
+                    .withBucketName(s3Config.getBucket())
+                    .withPrefix(s);
+            ObjectListing objectListing = s3Client.listObjects(listObjectsRequest);
+            return new VirtualFile[] {new S3VirtualFile(s3Client, objectListing.getObjectSummaries().get(0))};
+        }
+        if(isDirectory(s)){
+            if(!s.endsWith("/")) {
+                s += "/";
+            }
+            return convert(s);
+        }
+        String parentPath;
+        if(s.contains("/")) {
+            parentPath = s.substring(0, s.lastIndexOf('/'));
+        }
+        else {
+            parentPath = "";
+        }
+
+        if(!isDirectory(parentPath))
+            return null;
+        if(s.contains("*"))
+        {
+            return globSearch(parentPath, s);
+        }
+        return null;
+    }
+
+    private VirtualFile[] globSearch(String parentPath, String s) {
+        Vector<S3VirtualFile> s3VirtualFiles = new Vector<>();
```

Comment:
For just local usage an ArrayList is preferred to a Vector.

Reply:
Cool! I just read why is it so. Thanks!